mainchain-follower/db-sync-follower/src/candidates/datum/registered.rs

```diff
@@ -55,82 +56,64 @@ pub struct AdaBasedStaking {
-			let sidechain_signature = fields
-				.get(2)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| SidechainSignature(bytes.clone()))?;
-			let consumed_input = fields.get(3).and_then(decode_utxo_id_datum)?;
-			let _own_pkh = fields.get(4).and_then(|d| d.as_bytestring())?;
-			let aura_pub_key = fields
-				.get(5)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| AuraPublicKey(bytes.clone()))?;
-			let grandpa_pub_key = fields
-				.get(6)
-				.and_then(|d| d.as_bytestring())
-				.map(|bytes| GrandpaPublicKey(bytes.clone()))?;
-			Some(RegisterValidatorDatum::V0 {
-				stake_ownership,
-				sidechain_pub_key,
-				sidechain_signature,
-				consumed_input,
-				aura_pub_key,
-				grandpa_pub_key,
-			})
-		},
-
-		_ => None,
-	}
+pub fn decode_legacy_register_validator_datum(datum: PlutusData) -> Option<RegisterValidatorDatum> {
+	let fields = datum
+		.as_constr_plutus_data()
+		.filter(|datum| datum.alternative().is_zero())
+		.filter(|datum| datum.data().len() >= 7)?
+		.data();
+	let stake_ownership = decode_ada_based_staking_datum(fields.get(0))?;
+	let sidechain_pub_key = fields.get(1).as_bytes().map(SidechainPublicKey)?;
+	let sidechain_signature = fields.get(2).as_bytes().map(SidechainSignature)?;
+	let consumed_input = decode_utxo_id_datum(fields.get(3))?;
+	let _own_pkh = fields.get(4).as_bytes()?;
+	let aura_pub_key = fields.get(5).as_bytes().map(AuraPublicKey)?;
+	let grandpa_pub_key = fields.get(6).as_bytes().map(GrandpaPublicKey)?;
+	Some(RegisterValidatorDatum::V0 {
+		stake_ownership,
+		sidechain_pub_key,
+		sidechain_signature,
+		consumed_input,
+		aura_pub_key,
+		grandpa_pub_key,
+	})
 }
 
-fn decode_ada_based_staking_datum(datum: &Datum) -> Option<AdaBasedStaking> {
-	match datum {
-		ConstructorDatum { constructor: 0, fields } => match fields.first().zip(fields.get(1)) {
-			Some((ByteStringDatum(f0), ByteStringDatum(f1))) => {
-				let pub_key = TryFrom::try_from(f0.clone()).ok()?;
-				Some(AdaBasedStaking { pub_key, signature: MainchainSignature(f1.clone()) })
-			},
-			_ => None,
-		},
-		_ => None,
-	}
+fn decode_ada_based_staking_datum(datum: PlutusData) -> Option<AdaBasedStaking> {
+	let fields = datum
+		.as_constr_plutus_data()
+		.filter(|datum| datum.alternative().is_zero())
+		.filter(|datum| datum.data().len() >= 2)?
+		.data();
+	let pub_key = TryFrom::try_from(fields.get(0).as_bytes()?).ok()?;
+	let signature = MainchainSignature(fields.get(1).as_bytes()?);
+	Some(AdaBasedStaking { pub_key, signature })
 }
-fn decode_utxo_id_datum(datum: &Datum) -> Option<UtxoId> {
-	match datum {
-		ConstructorDatum { constructor: 0, fields } => match fields.first().zip(fields.get(1)) {
-			Some((f0, IntegerDatum(f1))) => {
-				let tx_hash = decode_tx_hash_datum(f0)?;
-				let index: u16 = TryFrom::try_from(f1.clone()).ok()?;
-				Some(UtxoId { tx_hash, index: UtxoIndex(index) })
-			},
-			_ => None,
-		},
-		_ => None,
-	}
+fn decode_utxo_id_datum(datum: PlutusData) -> Option<UtxoId> {
```

Comment:
Same here with a more explicit `if` and early exit

Reply:
I prefer the filtering style so I'll leave it as is.